src/libxtp/gwbse/sigma_spectral.cc

```diff
@@ -0,0 +1,225 @@
+      const Eigen::VectorXd& RPAEnergies = _rpa.getRPAInputEnergies();
+      const int numeigenvalues = _EigenSol._Omega.size();
+      
+      if (_HedinApprox) {
+
+        for (int s = 0; s < numeigenvalues; s++) {
+
+          double omega = _EigenSol._Omega(s);
+          Eigen::MatrixXd residues = CalcResidues(s);
+
+          for (int m = 0; m < _qptotal; m++) {
+
+            for (int n = 0; n < _qptotal; n++) {
+
+              if (m == n) {
+                continue; // Skip diagonal
+              }
+
+              result(m, n) += Equation48(m, n, omega, residues);
+
+            } // Energy level n
+          } // Energy level m
+        } // Eigenvalues/poles s
+
+      } else {
+
+        for (int s = 0; s < numeigenvalues; s++) {
+
+          double omega = _EigenSol._Omega(s);
+          Eigen::MatrixXd residues = CalcResidues(s);
+
+          for (int m = 0; m < _qptotal; m++) {
+
+            for (int n = 0; n < _qptotal; n++) {
+
+              if (m == n) {
+                continue; // Skip diagonal
+              }
+
+              // TODO: Take frequency or energy?
+              double w_m = RPAEnergies(m); // Frequency/energy
+              double w_n = RPAEnergies(n); // Frequency/energy
+
+              double result_m = Equation47(m, n, RPAEnergies, w_m, omega, residues);
+              double result_n = Equation47(m, n, RPAEnergies, w_n, omega, residues);
+
+              // (m|S(w)|n) = 0.5 * (m|S(e_m)|n) + 0.5 * (m|S(e_n)|n)
+              result(m, n) += 0.5 * (result_m + result_n);
+
+            } // Energy level n
+          } // Energy level m
+        } // Eigenvalues/poles s
+
+      }
+
+      return result;
+    }
+
+    Eigen::MatrixXd Sigma_Spectral::CalcResidues(int s) const {
+      Eigen::MatrixXd residues = Eigen::MatrixXd::Zero(_qptotal, _qptotal);
+      Eigen::VectorXd xpy = _EigenSol._XpY.col(s);
+      const int bsesize = (_homo - _qpmin + 1) * (_qpmax - (_homo + 1) + 1);
+      const int auxsize = _Mmn.auxsize(); // Size of gwbasis
+
+      for (int m = 0; m < _qptotal; m++) {
+        for (int n = 0; n < _qptotal; n++) {
+          for (int i_aux = 0; i_aux < auxsize; i_aux++) {
+
+            // Get three-center column for index (m, n)
+            VectorXfd tc_mn = _Mmn[m].col(i_aux);
+
+            for (int i = 0; i < bsesize; i++) {
+
+              int v = _vc2index.v(i);
+              int c = _vc2index.c(i);
+
+              // Get three-center column for index (v, c)
+              VectorXfd tc_vc = _Mmn[v].col(i_aux);
+
+              // Fill residues vector
+              residues(m, n) += tc_mn(n) * tc_vc(c) * xpy(i); // Eq. 45
+
+            } // Composite index i
+          } // Auxiliary basis function
+        } // Energy level n
+      } // Energy level m
+
+      return residues;
+    }
+
+    double Sigma_Spectral::Equation47(int m, int n, const Eigen::VectorXd& energies, double w, double omega, Eigen::MatrixXd& residues) const {
```

Comment:
name

Reply:
this function is not cool, 6 arguments is a bit much, are you sure you split your code like this